includes/gateway/class-omise-payment-creditcard.php

```diff
@@ -166,235 +169,192 @@ public function payment_fields() {
 
-						default:
-							// Default behaviour is, check if it paid first.
-							$success = Omise_Charge::is_paid( $charge );
+			if ( $this->omise_3ds ) {
+				$order->add_order_note(
+					sprintf(
+						__( 'Omise: Processing a 3-D Secure payment, redirecting buyer to %s', 'omise' ),
+						esc_url( $charge['authorize_uri'] )
+					)
+				);
 
-							// Then, check is authorized after if the first condition is false.
-							if ( ! $success )
-								$success = Omise_Charge::is_authorized( $charge );
+				return array(
+					'result'   => 'success',
+					'redirect' => $charge['authorize_uri'],
+				);
+			}
 
-							break;
+			switch ( $this->payment_action ) {
+				case self::PAYMENT_ACTION_AUTHORIZE:
+					$success = Omise_Charge::is_authorized( $charge );
+					if ( $success ) {
+						$order->add_order_note(
+							sprintf(
+								wp_kses(
+									__( 'Omise: Payment processing.<br/>An amount of %1$s %2$s has been authorized', 'omise' ),
+									array( 'br' => array() )
+								),
+								$order->get_total(),
+								$order->get_order_currency()
+							)
+						);
 					}
 
-					if ( ! $success ) {
-						throw new Exception( __( 'Note that your payment may have already been processed. Please contact our support team if you have any questions.', 'omise' ) );
+					break;
+
+				case self::PAYMENT_ACTION_AUTHORIZE_CAPTURE:
+					$success = Omise_Charge::is_paid( $charge );
+					if ( $success ) {
+						$order->payment_complete();
+						$order->add_order_note(
+							sprintf(
+								wp_kses(
+									__( 'Omise: Payment successful.<br/>An amount of %1$s %2$s has been paid', 'omise' ),
+									array( 'br' => array() )
+								),
+								$order->get_total(),
+								$order->get_order_currency()
+							)
+						);
 					}
 
-					// Remove cart
-					WC()->cart->empty_cart();
-					return array (
-						'result'   => 'success',
-						'redirect' => $this->get_return_url( $order )
-					);
-				}
-			} catch( Exception $e ) {
-				wc_add_notice(
-					sprintf(
-						wp_kses(
-							__( 'It seems we\'ve been unable to process your payment properly:<br/>%s', 'omise' ),
-							array( 'br' => array() )
-						),
-						$e->getMessage()
-					),
-					'error'
-				);
+					break;
 
-				$order->add_order_note(
-					sprintf(
-						__( 'Omise: Payment failed, %s', 'omise' ),
-						$e->getMessage()
-					)
-				);
+				default:
+					// Default behaviour is, check if it paid first.
+					$success = Omise_Charge::is_paid( $charge );
 
-				return;
+					// Then, check is authorized after if the first condition is false.
+					if ( ! $success )
+						$success = Omise_Charge::is_authorized( $charge );
+
+					break;
+			}
+
+			if ( ! $success ) {
+				return $this->payment_failed( __( 'Note that your payment may have already been processed. Please contact our support team if you have any questions.', 'omise' ) );
 			}
+
+			// Remove cart
+			WC()->cart->empty_cart();
+			return array (
+				'result'   => 'success',
```

Comment:
Perhaps 'success' shouldn't be hardcoded?

Reply:
@jonrandy maybe, any idea on its constant name?

Reply:
Looks like something that is maybe already in WC? Is this return in a particular format that WC is expecting?

Reply:
@jonrandy yup, a payment provider who implemented this `process_payment` is expected to return `result` and `redirect` in an array format back.

Here is a usage-reference in WC class that executed `process_payment` method.
<img width="1055" alt="Screen Shot 2562-04-26 at 15 09 58" src="https://user-images.githubusercontent.com/2154669/56793638-a5299800-6836-11e9-8d54-5b0365f00f92.png">

Reply:
I guess if they don't have a constant for it, we can leave it